chatcommunicate.py

```diff
@@ -197,459 +196,41 @@ def handle_commands(content_lower, message_parts, ev_room, ev_room_name, ev_user
-        se_site = message_parts[2].replace('-', '')
-        response = remove_from_notification_list(user_id, chat_site, room_id, se_site)
-        if response:
-            if quiet_action:
-                return None
-            return "I will no longer ping you if I report a post on `{site_name}`, in room `{room_id}` on `chat.{chat_domain}`".format(site_name=se_site, room_id=room_id, chat_domain=chat_site)
-        return "That configuration doesn't exist."
-    if content_lower.startswith("!!/willibenotified"):
-        if len(message_parts) != 3:
-            return False, "2 arguments expected"
-        user_id = int(ev_user_id)
-        chat_site = wrap2.host
-        room_id = message_parts[1]
-        if not room_id.isdigit():
-            return False, "Room ID is invalid"
-
-        room_id = int(room_id)
-        se_site = message_parts[2]
-        will_be_notified = will_i_be_notified(user_id, chat_site, room_id, se_site)
-        if will_be_notified:
-            return "Yes, you will be notified for that site in that room."
-
-        return "No, you won't be notified for that site in that room."
-    if content_lower.startswith("!!/allnotificationsites"):
-        if len(message_parts) != 2:
-            return False, "1 argument expected"
-        user_id = int(ev_user_id)
-        chat_site = wrap2.host
-        room_id = message_parts[1]
-        if not room_id.isdigit():
-            return False, "Room ID is invalid."
-        sites = get_all_notification_sites(user_id, chat_site, room_id)
-        if len(sites) == 0:
-            return "You won't get notified for any sites in that room."
 
-        return "You will get notified for these sites:\r\n" + ", ".join(sites)
-    return False, None  # Unrecognized command, can be edited later.
+        subcommand_parameters = {
+            'msg_content': msg_content,
+            'ev_room': ev_room,
+            'ev_room_name': ev_room_name,
+            'ev_user_id': ev_user_id,
+            'ev_user_name': ev_user_name,
+            'message_url': message_url,
+            'msg': msg,
+            'post_site_id': post_site_id,
+            'post_type': post_type,
+            'post_url': post_url,
+            'quiet_action': quiet_action,
+            'second_part_lower': second_part_lower,
+            'wrap2': wrap2,
+        }
+
+        try:
+            return subcmds[second_part_lower](**subcommand_parameters)
+        except KeyError:
+            pass    # Not a valid subcommand, skip
+
+    # Process additional commands
+    command_parameters = {
+        'content': content,
+        'content_lower': content_lower,
+        'ev_room': ev_room,
+        'ev_room_name': ev_room_name,
+        'ev_user_id': ev_user_id,
+        'ev_user_name': ev_user_name,
+        'message_parts': message_parts,
+        'message_url': message_url,
+        'wrap2': wrap2,
+    }
+    try:
+        return cmds[command](**command_parameters)
+    except KeyError:
```

Comment:
Sure, I totally get the concept of EAFP and would normally be like "yeah, that makes sense" except in this case _this is hiding potential bugs in the commands_.

I roughly estimated the number of times a dictionary lookup is done in `chatcommands.py` by doing a quick search for `['`, which appears to have under 10 false positives. I got 185 hits for that, which means there **185 chances that a command raises a `KeyError`**, which would be ignored and the user would be notified that they used an unrecognized command.

So, sure. Here's the EAFP way of doing this, such that you don't have that problem.

```python
try:
    comm = cmds[command]
except KeyError:
    return False, None
else:
    return comm(**command_parameters)
```

And, for those who are interested, the LBYL way of doing it

```python
if command not in cmds:
    return False, None

return cmds[command](**command_parameters)
```